quickjs/src/jniMain/kotlin/app/cash/quickjs/Zipline.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.quickjs
+
+import app.cash.quickjs.internal.bridge.InboundService
+import app.cash.quickjs.internal.bridge.OutboundClientFactory
+import java.util.concurrent.Executors
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.launch
+
+actual abstract class Zipline {
+  actual abstract val engineVersion: String
+
+  abstract val quickJs: QuickJs
+
+  abstract val dispatcher: CoroutineDispatcher
+
+  actual fun <T : Any> get(name: String, jsAdapter: JsAdapter): T {
+    error("unexpected call to Zipline.get: is the Zipline plugin configured?")
+  }
+
+  @PublishedApi
+  internal abstract fun <T : Any> get(
+    name: String,
+    outboundClientFactory: OutboundClientFactory<T>
+  ): T
+
+  actual fun <T : Any> set(name: String, jsAdapter: JsAdapter, instance: T) {
+    error("unexpected call to Zipline.set: is the Zipline plugin configured?")
+  }
+
+  @PublishedApi
+  internal abstract fun set(name: String, service: InboundService<*>)
+
+  companion object {
+    fun create(
+      dispatcher: CoroutineDispatcher = Executors.newSingleThreadExecutor().asCoroutineDispatcher(),
+    ): Zipline {
+      // TODO(jwilson): avoid runBlocking if I’m already on our dispatcher??
+      return kotlinx.coroutines.runBlocking(dispatcher) {
```

Comment:
I don't see why it's needed at all. No one can interact with the instance until it's returned so we're safe to create and initialize on the caller thread.

Reply:
Yeah, good call. Done.